lua/entities/ent_jack_gmod_ezsolargenerator.lua

```diff
@@ -0,0 +1,247 @@
+        self:SetSolid(SOLID_VPHYSICS)
+        self:DrawShadow(true)
+        self:SetUseType(SIMPLE_USE)
+        local phys = self:GetPhysicsObject()
+        if phys:IsValid() then
+            phys:Wake()
+            phys:SetMass(200)
+        end
+        self:SetState(STATE_OFF)
+        self:SetProgress(0)
+        self.NextUse = 0
+        local mapName = game.GetMap()
+        if(string.find(mapName, "_night") or string.find(mapName, "night_"))then self.NightMap=true end
+    end
+    
+
+   function ENT:Use(activator)
+		local State=self:GetState()
+		local OldOwner=self.Owner
+		JMod.Owner(self,activator)
+		JMod.Colorify(self)
+		if(IsValid(self.Owner))then
+			if(OldOwner~=self.Owner)then -- if owner changed then reset team color
+				JMod.Colorify(self)
+			end
+		end
+		if(State==STATE_BROKEN)then
+			JMod.Hint(activator,"destroyed",self)
+			return
+		elseif(State==STATE_OFF)then
+			self:TurnOn()
+		elseif(State==STATE_ON)then
+            self:ProducePower()
+			self:TurnOff()
+		end
+	end
+    
+    function ENT:SpawnEffect(pos)
+        local effectdata=EffectData()
+        effectdata:SetOrigin(pos)
+        effectdata:SetNormal((VectorRand()+Vector(0,0,1)):GetNormalized())
+        effectdata:SetMagnitude(math.Rand(5,10))
+        effectdata:SetScale(math.Rand(.5,1.5))
+        effectdata:SetRadius(math.Rand(2,4))
+        util.Effect("Sparks", effectdata)
+        self:EmitSound("items/suitchargeok1.wav", 80, 120)
+    end
+    
+    function ENT:ProducePower()
+        local SelfPos,Up,Forward,Right = self:GetPos(),self:GetUp(),self:GetForward(),self:GetRight()
+        local amt = math.min(math.floor(self:GetProgress()), self.MaxPower)
+
+        if amt <= 0 then return end
+        
+        local pos = SelfPos + Forward*15 - Up*25 - Right*2
+        for _, ent in pairs(ents.FindInSphere(pos, 100)) do
+            --print(ent, ent.GetResourceType and ent:GetResourceType())
+            if ((ent:GetClass() == "ent_jack_gmod_ezcrate") and (ent:GetResourceType() == "generic" 
+            or ent:GetResourceType() == "power") and (ent:GetResource() + amt <= ent.MaxResource)) then
+                    
+                if ent:GetResourceType() == "generic" then
+                    ent:ApplySupplyType("power")
+                end
+                    
+                ent:SetResource(math.min(ent:GetResource() + amt, ent.MaxResource))
+                self:SetProgress(self:GetProgress() - amt)
+                self:SpawnEffect(pos)
+                return
+            end
+        end
+        JMod.MachineSpawnResource(self, "power", amt, self:WorldToLocal(pos), Angle(-90, 0, 0), Up*-300)
+        self:SetProgress(self:GetProgress() - amt)
+        self:SpawnEffect(pos)
+    end
+
+    function ENT:CheckSky()
+        local HitAmount = 0
+        for i = 1, 10 do
+            for j = 1, 10 do
+                local StartPos = self:LocalToWorld(Vector(-5 + j*1, -100 + i*25, 10 + j*7.5))
```

Comment:
the math gives me the big WHEEEEE

Reply:
It took a minute, but I got it working properly in the end.